Adding support for Thurlby 1905a DMM

m-charlton opened the pull request:
The Thurlby 1905a 'intelligent' digital multimeter outputs its readings via an RS-232 serial port. Submitted code enables the user to read from this port.

Tested against a real instrument. 

First contribution to this project, so seeking feedback on any oversights/errors made

m-charlton committed:
Adding support for Thurlby 1905a DMM

The Thurlby 1905a 'intelligent' digital multimeter outputs its readings
via an RS-232 serial port. Submitted code enables the user to read from
this port.

Tested against a real instrument.

m-charlton committed:
Correction to methods docstring

Methods should use 'Control', 'Measure', 'Set' or 'Get' in
method docstrings.

m-charlton committed:
Correcting lazy cut'n paste error

Copied an existing index file as the template for a new index
but forgot to change some of the text.

BenediktBurger commented on docs/api/instruments/tti/index.rst:
```diff
@@ -0,0 +1,13 @@
+###########################
+
+
+This section contains specific documentation on the Thurlby Thandar Instruments (TTi), Thurlby and Thandar instruments that are implemented. If you are interested in an instrument not included, please consider :doc:`adding the instrument </dev/adding_instruments/index>`.
```
That does not sound good: `Thurlby Thandar Instruments (TTi), Thurlby and Thandar instruments`. I do not understand, what you want to say.

m-charlton replied:
Thurlby & Thandar were two separate companies. At some point they merged to form a new company "Thurlby Thandar Instruments" (TTi).

How about I rename the directory `thurlby` and if, in the future, someone wants to add a Thandar or TTi instrument then they can create a directory called `thandar` or `tti`?

BenediktBurger replied:
"tti" is fine.
As I did not know that history, I got confused by that phrasing.
You could use something like, `Thurlby Thandar Instruments (TTi) or one of their predecessors Thurlby Instruments and Thandard Instruments`.

BenediktBurger commented on docs/api/instruments/tti/thurlby1905a.rst:
```diff
@@ -0,0 +1,24 @@
+
+.. contents::
+
+**********************************************
+General Information
+**********************************************
+This instrument only provides measurement output via an RS-232 serial port.
+The serial port can only be read from, not written to.
+
+Earlier models had a Baud rate of 2400, later ones 9600.
+
+Initialization of the Instrument
+====================================
+
+.. code-block:: python
+
+    from pymeasure.instruments.tti import Thurlby1905a
+
+    dmm = Thurlby1905a("ASRL/dev/ttyS0::INSTR")
+    output = dmm.read_measurement
```
Put this information into the device docstring itself.
As you will import that device, the docstring will be visible.

m-charlton replied:
Will do.

BenediktBurger commented on docs/api/instruments/tti/thurlby1905a.rst:
```diff
@@ -0,0 +1,24 @@
+############################################
+Thurlby 1905a intelligent digital multimeter
+############################################
+.. currentmodule:: pymeasure.instruments.tti.thurlby1905a
```
We use `..autoclass` to import classes into the documentation, see any other instrument file.

BenediktBurger commented on pymeasure/instruments/tti/thurlby1905a.py:
```diff
@@ -0,0 +1,110 @@
+# THE SOFTWARE.
+#
+
+import logging
+
+from pymeasure.errors import Error, RangeError
+from pymeasure.instruments import Instrument
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
```
If you do not log anything, you do not need to import it and create `log`.

BenediktBurger commented on pymeasure/instruments/tti/thurlby1905a.py:
```diff
@@ -0,0 +1,110 @@
+    """
+
+    def __init__(self, adapter, name="Thurlby 1905a", **kwargs):
+        kwargs.setdefault("read_termination", "\r")
```
Can you change the read_termination? If not, just write it into the super.init call.

m-charlton replied:
Can guarantee that `read_termination`, for this version at least, is '\r'. Will do as suggested.

BenediktBurger commented on pymeasure/instruments/tti/thurlby1905a.py:
```diff
@@ -0,0 +1,110 @@
+        #   'M  Cd  00 '
+
+        if len(measurement) != 10:
+            raise ValueError("Measurement '%s' is not expected 10 bytes long." % measurement)
```
f-strings are easier to read and more powerful:

```suggestion
            raise ValueError(f"Measurement '{measurement}' is not expected 10 bytes long.")
```

BenediktBurger commented on pymeasure/instruments/tti/thurlby1905a.py:
```diff
@@ -0,0 +1,110 @@
+        )
+
+    @staticmethod
+    def _translate(measurement):
```
maybe "parse"?

m-charlton replied:
Will do.

BenediktBurger commented on pymeasure/instruments/tti/thurlby1905a.py:
```diff
@@ -0,0 +1,110 @@
+
+    @staticmethod
+    def _translate(measurement):
+        # Easier to test if a static method
+        #
+        # There are two types of measurements. A valid reading 'R' or a message 'M'.
+        # All measurements, excluding termination characters, are 10 bytes long.
+        #
+        # Examples:
+        #  Valid reading of -0.0
+        #   'R- 000.00 '
+        #
+        #  Valid reading of 997.628
+        #   'R  997.628'
+        #
+        # A message is anything other than a reading:
+        #
+        #  Over Range
+        #   'M   OR    '
+        #
+        #  An error
+        #   'M   ERROR '
+        #
+        # The instruments software version, issued, briefly, at power-up:
+        #   'M  Cd  00 '
```
Why don't you write a normal docstring?

The static method comment can stay a comment as it is for implementation.

BenediktBurger commented on tests/instruments/tti/test_thurlby1905a.py:
Thanks for adding tests. Did you look at our protocol tests: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/tests.html#protocol-tests ? They are quite powerful.

BenediktBurger reviewed (changes requested):
Thanks for your contribution. As a new contributor, you may add yourself to the authors.txt file.

Here are a few comments.

m-charlton commented on pymeasure/instruments/tti/thurlby1905a.py:
```diff
@@ -0,0 +1,110 @@
+
+        raise Error("Unknown type of measurement '%s'." % measurement)
+
+    @property
+    def read_measurement(self):
```
That makes more sense.

m-charlton committed:
Changes after review #1345

Changes after code review:

- tti -> thurlby directory
- moved documentation to python code docstrings
- removed extraneous code
- minor refactoring
- added protocol tests

BenediktBurger commented on pymeasure/instruments/thurlby/thurlby1905a.py:
```diff
@@ -0,0 +1,126 @@
+
+    """
+
+    def __init__(self, adapter, name="Thurlby 1905a", **kwargs):
+        super().__init__(
+            adapter,
+            name,
+            includeSCPI=False,
+            read_termination="\r",
+            **kwargs,
+        )
```
As different devices could have different baud rates, I'd suggest to add it to the signature:

```suggestion
    def __init__(self, adapter, name="Thurlby 1905a", baud_rate=2400, **kwargs):
        super().__init__(
            adapter,
            name,
            includeSCPI=False,
            read_termination="\r",
            baud_rate=baud_rate,
            **kwargs,
        )
```

Change the default to 9600, if you prefer.

BenediktBurger reviewed (approved):
Looks good, thanks for the changes and for your patience @m-charlton .

Here a few small comments.
